addons/website_sale/controllers/main.py

```diff
@@ -1102,6 +1108,129 @@ def address(self, **kw):
+
+        # In a non-express flow, `sale_last_order_id` would be added in the session before the
+        # payment. As we skip all the steps with the express checkout, `sale_last_order_id` must be
+        # assigned to ensure the right behavior from `shop_payment_confirmation()`.
+        request.session['sale_last_order_id'] = order.id
+
+        return order.partner_id.id
+
+    def _include_country_and_state_in_address(self, address):
+        """ This function is used to include country_id and state_id in address.
+
+        Fetch country and state and include the records in address. The object is included to
+        simplify the comparison of addresses.
+
+        :param dict address: An address with country and state defined in ISO 3166.
+        :return None:
+        """
+        country = request.env["res.country"].search([
+            ('code', '=', address['country']),
+        ], limit=1)
+        state = request.env["res.country.state"].search([
+            ('code', '=', address['state']),
+        ], limit=1)
+        address['country_id'] = country
+        address['state_id'] = state
+        del address['country']
+        del address['state']
+
+    def _convert_country_and_state_records_into_int(self, address):
+        """ This function is used to convert into integers country_id and state_id.
+
+        :param dict address: An address with country and state defined in ISO 3166.
+        :return None:
+        """
+        address['country_id'] = address['country_id'].id
+        address['state_id'] = address['state_id'].id
+
+    def _create_or_edit_partner(self, values, edit=False):
+        """ Create or update a partner
+
+        To create a partner, this controller usually calls `values_preprocess()`, then
+        `checkout_form_validate()`, then `values_postprocess()` and finally `_checkout_form_save()`.
+        Since these methods are very specific to the checkout form, this method makes it possible to
+        create  a partner for more specific flows like express payment, which does not require all
+        the checks carried out by the previous methods. Parts of code in this method come from those.
+
+        :param dict values: The values needed to create the partner or to edit the partner.
+        :param bool edit: Whether edit an existing partner or create one, defaults to False.
+        :return int: The id of the partner created or edited
+        """
+        if request.website.specific_user_account:
+            values['website_id'] = request.website.id
+
+        lang = request.lang.code if request.lang.code in request.website.mapped('language_ids.code') else None
+        if lang:
+            values['lang'] = lang
+
+        Partner = request.env['res.partner']
+        if edit:
+            partner_id = int(values.get('partner_id', 0))
+            if partner_id:
+                del values['partner_id']  # Can't edit the partner_id
+                Partner.browse(partner_id).sudo().write(values)
```

Comment:
I am not sure if it is the case here, but some fields on Partner should be prevented from being written to in some cases, see https://github.com/odoo/odoo/blob/master/addons/account/controllers/portal.py#L130

Reply:
Note: it is possible to do with some other `website_sale`'s routes. I got aware of that issue only recently.

Reply:
Thanks for pointing that out. I hadn't thought of it that way. I'll change the code to sanitize the fields I'm writing on.